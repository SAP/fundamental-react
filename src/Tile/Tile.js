import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import TileContent from './_TileContent';
import TileFooter from './_TileFooter';
import TileHeader from './_TileHeader';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/tile.css';

const classnames = classnamesBind.bind(styles);

/** A **Tile** can be used to represent an app similar to the SAP Fiori launchpad home page.
 * They can display different types of content, which are based on the data supplied by the app.
 * They can contain an icon, a title, an informative text, KPIs, counters and charts. */

const Tile = React.forwardRef(({
    disabled,
    backgroundImage,
    children,
    className,
    cssNamespace,
    isDouble,
    onClick,
    size,
    ...rest
}, ref) => {

    const tileClasses = classnames(
        `${cssNamespace}-tile`,
        {
            [`${cssNamespace}-tile--${size}`]: size === 's',
            [`${cssNamespace}-tile--double`]: isDouble,
            'is-disabled': disabled
        },
        className
    );

    return (
        <div
            {...rest}
            className={tileClasses}
            onClick={onClick}
            ref={ref}
            role='button'
            tabIndex='0'>
            {children}
        </div>
    );
});

Tile.displayName = 'Tile';

Tile.propTypes = {
    /** Callback function when user clicks on the component*/
    onClick: PropTypes.func.isRequired,
    /** URL of the background image for product tile */
    backgroundImage: PropTypes.string,
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to mark component as disabled and make it non-interactive */
    disabled: PropTypes.bool,
    /** Set to **true** to increase tile dimensions to 2x1 */
    isDouble: PropTypes.bool,
    /** Default tiles are designed for screens larger than 374px. For smaller screens use 's' */
    size: PropTypes.oneOf(['s'])
};

Tile.Content = TileContent;
Tile.Header = TileHeader;
Tile.Footer = TileFooter;

export default withStyles(Tile);
