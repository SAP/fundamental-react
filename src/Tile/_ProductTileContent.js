import React from 'react';
import TileContent from './_TileContent';

const ProductTileContent = props => {
    return (<TileContent productTile {...props} />);
};

ProductTileContent.displayName = 'ProductTile.Content';

export default ProductTileContent;
