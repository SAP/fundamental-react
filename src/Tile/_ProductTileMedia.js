import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const ProductTileMedia = props => {
    const { image, className, ...rest } = props;

    const tileProductMediaClasses = classnames(
        'fd-product-tile__media',
        className
    );

    return (
        <div
            {...rest}
            className={tileProductMediaClasses}
            style={{ backgroundImage: 'url(' + image + ')' }} />
    );
};

ProductTileMedia.displayName = 'ProductTileMedia';

ProductTileMedia.propTypes = {
    image: PropTypes.string.isRequired,
    className: PropTypes.string
};

ProductTileMedia.propDescriptions = {
    image: 'URL of the image.'
};

export default ProductTileMedia;
