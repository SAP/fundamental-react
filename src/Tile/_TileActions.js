import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const TileActions = props => {
    const { children, className, ...rest } = props;

    const tileActionsClasses = classnames(
        'fd-tile__actions',
        className
    );

    return <div {...rest} className={tileActionsClasses}>{children}</div>;
};

TileActions.displayName = 'Tile.Actions';

TileActions.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string
};

export default TileActions;
