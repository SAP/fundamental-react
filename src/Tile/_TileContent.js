import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/tile.css';

const classnames = classnamesBind.bind(styles);

const TileContent = props => {
    const { children, className, cssNamespace, twoColumns, ...rest } = props;

    const tileContentClasses = classnames(
        `${cssNamespace}-tile__content`,
        {
            [`${cssNamespace}-tile__content--2-col`]: twoColumns
        },
        className
    );

    return (
        <div {...rest} className={tileContentClasses}>
            {twoColumns ? React.Children.toArray(children).map(child => {
                return React.cloneElement(child, {
                    className: classnames(child.className, `${cssNamespace}-tile__section`)
                });
            }) : children }
        </div>
    );
};

TileContent.displayName = 'Tile.Content';

TileContent.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to split TileContent into two columns with a 0.25rem vertical padding.
     * Any children must be wrapped in 2 top level `<div>` elements.
    */
    twoColumns: PropTypes.bool
};

export default withStyles(TileContent);
