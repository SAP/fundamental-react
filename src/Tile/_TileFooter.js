import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/tile.css';

const classnames = classnamesBind.bind(styles);

const TileFooter = props => {
    const { children, className, cssNamespace, ...rest } = props;

    const tileFooterClasses = classnames(
        `${cssNamespace}-tile__footer`,
        className
    );

    return (
        <div {...rest} className={tileFooterClasses}>
            {children}
        </div>);
};

TileFooter.displayName = 'Tile.Footer';

TileFooter.propTypes = {
    /** Node(s) to render inside the header element */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string
};

export default withStyles(TileFooter);
