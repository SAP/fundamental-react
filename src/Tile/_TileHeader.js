import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/tile.css';

const classnames = classnamesBind.bind(styles);

const TileHeader = props => {
    const { children, className, cssNamespace, subtitle, subtitleProps, titleProps, ...rest } = props;

    const tileHeaderClasses = classnames(
        `${cssNamespace}-tile__header`,
        className
    );

    const tileTitleClasses = classnames(
        `${cssNamespace}-tile__title`,
        titleProps?.className
    );

    const tileSubtitleClasses = classnames(
        `${cssNamespace}-tile__subtitle`,
        subtitleProps?.className
    );

    return (
        <div {...rest} className={tileHeaderClasses}>
            <div {...titleProps} className={tileTitleClasses}>{children}</div>
            {subtitle && (<div {...subtitleProps} className={tileSubtitleClasses}>{subtitle}</div>)}
        </div>);
};

TileHeader.displayName = 'Tile.Header';

TileHeader.propTypes = {
    /** Node(s) to render inside the header element */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Node(s) to render inside the optional subtitle element */
    subtitle: PropTypes.node,
    /** Additional props to be spread to the subtitle `<div>` element */
    subtitleProps: PropTypes.object,
    /** Additional props to be spread to the title `<div>` element */
    titleProps: PropTypes.object
};

export default withStyles(TileHeader);
