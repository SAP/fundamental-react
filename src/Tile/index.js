export { default as Tile } from './Tile.js';
export { default as TileGrid } from './TileGrid.js';
