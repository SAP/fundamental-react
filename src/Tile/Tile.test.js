import { mount } from 'enzyme';
import React from 'react';
import Tile from './Tile';

describe('<Tile />', () => {
    const handleClick = jest.fn();
    const activeTile = (
        <Tile onClick={handleClick}>
            <Tile.Header>Test Header</Tile.Header>
            <Tile.Content>Test Content</Tile.Content>
        </Tile>
    );

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Tile component', () => {
            const element = mount(<Tile data-sample='Sample' onClick={handleClick} />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <Tile onClick={handleClick} ref={ref} />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('DIV');
        expect(ref.current.className).toEqual('fd-tile');
    });

    test('onClick handler should be tile', () => {
        const component = mount(activeTile);
        component
            .find('div.fd-tile')
            .simulate('click', { target: { text: '4' }, stopPropagation: () => {} });
        expect(handleClick).toHaveBeenCalled();
    });
});
