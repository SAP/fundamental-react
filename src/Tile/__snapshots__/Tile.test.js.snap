// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> create tile component 1`] = `
<div
  className="fd-tile blue"
>
  <div
    className="fd-tile__content red"
  >
    <h2
      className="fd-tile__title"
    >
      Tile Title
    </h2>
    <p>
      Tile Description
    </p>
  </div>
</div>
`;

exports[`<Tile /> create tile component 2`] = `
<div
  className="fd-tile is-disabled fd-has-grid-column-span-3 fd-has-background-color-background-8"
>
  <div
    className="fd-tile__content"
  >
    <h2
      className="fd-tile__title"
    >
      Tile Title
    </h2>
    <p>
      Tile Description
    </p>
  </div>
</div>
`;

exports[`<Tile /> create tile component 3`] = `
<div
  className="fd-tile"
  isButton={true}
>
  <div
    className="fd-tile__media green"
  >
    <span
      className="fd-identifier--m sap-icon--home fd-has-background-color-accent-3"
      role="presentation"
    />
  </div>
  <div
    className="fd-tile__content"
  >
    <h2
      className="fd-tile__title"
    >
      Tile Title
    </h2>
    <p>
      Tile Description
    </p>
  </div>
</div>
`;

exports[`<Tile /> create tile component 4`] = `
<div
  className="fd-tile"
>
  <div
    className="fd-tile__content"
  >
    <h2
      className="fd-tile__title"
    >
      Tile Title
    </h2>
  </div>
  <div
    className="fd-tile__actions yellow"
  >
    <div
      className="fd-popover"
    >
      <div
        aria-expanded={false}
        className="fd-popover__control"
        onClick={[Function]}
      >
        <button
          className="fd-button fd-button--standard sap-icon--vertical-grip"
          disabled={false}
          selected={false}
        />
      </div>
      <div
        aria-hidden={true}
        className="fd-popover__body"
      >
        <nav
          className="fd-menu"
        >
          <ul
            className="fd-menu__list"
          >
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 1
              </a>
            </li>
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 2
              </a>
            </li>
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 3
              </a>
            </li>
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 4
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Tile /> create tile component 5`] = `
<div
  className="fd-tile"
  isButton={true}
>
  <div
    className="fd-tile__media"
  >
    <span
      className="fd-identifier--m sap-icon--home fd-has-background-color-accent-3"
      role="presentation"
    />
  </div>
  <div
    className="fd-tile__content"
  >
    <h2
      className="fd-tile__title"
    >
      Tile Title
    </h2>
    <p>
      Tile Description
    </p>
  </div>
</div>
`;

exports[`<Tile /> create tile component 6`] = `
<div
  className="fd-tile"
>
  <div
    className="fd-tile__content"
  >
    <h2
      className="fd-tile__title"
    >
      Tile Title
    </h2>
  </div>
  <div
    className="fd-tile__actions"
  >
    <div
      className="fd-popover"
    >
      <div
        aria-expanded={false}
        className="fd-popover__control"
        onClick={[Function]}
      >
        <button
          className="fd-button fd-button--standard sap-icon--vertical-grip"
          disabled={false}
          selected={false}
        />
      </div>
      <div
        aria-hidden={true}
        className="fd-popover__body"
      >
        <nav
          className="fd-menu"
        >
          <ul
            className="fd-menu__list"
          >
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 1
              </a>
            </li>
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 2
              </a>
            </li>
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 3
              </a>
            </li>
            <li>
              <a
                className="fd-menu__item"
                href="/"
              >
                Option 4
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Tile /> create tile component 7`] = `
<div
  className="fd-product-tile pink"
  isButton={true}
>
  <div
    className="fd-product-tile__media"
    style={
      Object {
        "backgroundImage": "url(https://techne.yaas.io/images/product-thumbnail-wide.png)",
      }
    }
  />
  <div
    className="fd-product-tile__content"
  >
    <h2
      className="fd-product-tile__title"
    >
      Tile Title
    </h2>
    <p>
      Tile Description
    </p>
  </div>
</div>
`;

exports[`<Tile /> create tile component 8`] = `
<div
  className="fd-product-tile is-disabled"
>
  <div
    className="fd-product-tile__media blue"
    style={
      Object {
        "backgroundImage": "url(https://techne.yaas.io/images/product-thumbnail-wide.png)",
      }
    }
  />
  <div
    className="fd-product-tile__content blue"
  >
    <h2
      className="fd-product-tile__title"
    >
      Tile Title
    </h2>
    <p>
      Tile Description
    </p>
  </div>
</div>
`;

exports[`<Tile /> create tile component 9`] = `
<div
  className="fd-tile-grid fd-tile-grid--3col blue"
>
  <div
    className="fd-tile fd-has-grid-row-span-2 fd-has-background-color-accent-7"
  >
    <div
      className="fd-tile__content"
    >
      <h2
        className="fd-tile__title"
      >
        Tile Title
      </h2>
      <p>
        Tile Description
      </p>
    </div>
  </div>
</div>
`;

exports[`<Tile /> create tile component 10`] = `
<div
  className="fd-tile-grid fd-tile-grid--4col"
>
  <div
    className="fd-tile fd-has-grid-row-span-2 fd-has-background-color-accent-7"
  >
    <div
      className="fd-tile__content"
    >
      <h2
        className="fd-tile__title"
      >
        Tile Title
      </h2>
      <p>
        Tile Description
      </p>
    </div>
  </div>
</div>
`;
